Extended Rearrangement

RLee64 opened the pull request:
## Overview
In #1 we resolved the rearrangement example seen in #1354 for the trivial x = y + z case. However, we want to expand upon this functionality in order to support more operations. As such, this pull request focuses on growing the existing level of support for equation rearrangement, expanding functionality to other use cases. That being said, this implementation does not fully cover all libcellml supported types, but we hope to expand upon this in an additional pull request.

Here we also begin to build a set of proper test cases for us to ensure SymEngine's rearrangement functionality is behaving as we intend.

## Changes
- Added test cases for equation rearrangement (originally written by GPT, but I've adjusted files to fix errors and hopefully cover more edge cases)
- Parsing from SymEngine back to AST now supports SymEngine expressions with 3+ children
- Rearrangement support for addition and subtraction
- Rearrangement support for unary +/-
- Rearrangement support for multiplication and division
- Rearrangement support for basic trigonometric rearrangement (when the unknown variable is not inside a trig function)
- Rearrangement support for various constants (i.e. e, pi, inf)
- Rearrangement support for different number 'types' (i.e. doubles and scientific notation)
- Rearrangement support for polynomials with one **real** solution
- In order to achieve polynomial rearrangement support, the method ``isSymEngineExpressionComplex`` was introduced to recursively look for complex constituents. This is consequently used to remove complex equations from our solution set.

## Notes
- Since SymEngine occasionally throws exceptions when the system is unable to rearrange an equation (e.g. when needing to rearrange tan to atan or having to work with fractional polynomials), I'm guessing we'll probably want to catch these errors at some point to ensure that these errors don't terminate code execution just because rearrangement wasn't successful. With that in mind, I haven't done it yet since getting symengine errors is useful for discovering the library's limitations.
- Similar to the previous PR, we're once again failing some existing test cases. At a quick glance, it seems that the unordered algebraic system is running into an issue with variables now being underconstrained and the hodgkin huxley model tests now have some equations interpreted as algebraic. I'll have a deeper look into these to get a better idea of what's going on.
<div style="display:block; margin:auto;">
<img width="1378" height="276" alt="image" src="https://github.com/user-attachments/assets/8c67e704-19c5-49a4-aab9-3a5481c44b54" />
<div />

- Not all of the new test cases currently pass, but this PR is probably big enough as is, so I'm hoping to finish them off in a third branch that wraps up rearrangement (at least for the time being).

<div style="display:block; margin:auto;">
<img width="500" height="150" alt="image" src="https://github.com/user-attachments/assets/651bae0a-84f0-45c0-8ece-96dbe0e0b105" />
<div />

RLee64 committed:
Added rearrangement test suite

Mostly GPT generated test cases

RLee64 committed:
Create new test cases for rearrangement

RLee64 committed:
Remove old rearrangement test case

Test case has been superseded by the newly added test suite to more robustly test for equation rearrangement

RLee64 committed:
Make SymEngine -> AST parsing work with 3+ children

Previous system assumed that symengine expressions contained only 0-2 children, but in reality it could be any whole number. This adjusts the existing conversion logic to account for this.

RLee64 committed:
Add/Sub symengine parsing support

Can now handle unary plus, minus, unary minus, and CN types from AST (which means code now passes Analyser.rearrangeAdditiveEquations)

RLee64 committed:
Fixed unarranged multiplication cellml file

Had an error at equation 4

Also added brackets to equation comments to make things more clear

RLee64 committed:
Updated unarranged addition cellml file

Made it so that we also now consider mathml unary +/- operators

RLee64 committed:
Mult/div symengine parsing support

Now passes Analyser.rearrangeMultiplicativeEquations

RLee64 committed:
Improved unarranged trig test cases

Now covers a greater variety of cases

RLee64 committed:
Simple trig rearrangement support

Symengine is not capable of rearranging trig functions into inverse trig functions (e.g. tan(x) = y cannot become x = atan(y))

RLee64 committed:
Updated unarranged constants cellml file

Now has better Eq 1 test case and more consistent variable naming

RLee64 committed:
Added support for various numbers and constants

Now supports
- Decimal point values (i.e. doubles)
- Constants (E, pi, and inf)

RLee64 committed:
Fix unarranged polynomial test cases

We need to ensure that there can only be one real solution for each

RLee64 committed:
Added support for rearranging polynomials

Some specific things had to be added to handle this
- CN representing integers must be appropriately converted since symengine requires exponentials to be whole numbers
- We need to filter out real from non-real solutions since we assume users will want to ignore the complex domain in most cases

RLee64 committed:
Update unary +/- parsing

Now assumes left child will always be the non-null component

agarny commented on src/analyser.cpp:
Why limit ourselves to what you have here? Can we not support the conversion of **_everything_** CellML supports and leave it to SymEngine to do whatever it can with it? In other words, I would convert everything that  is listed [here](https://github.com/RLee64/libcellml/blob/main/src/api/libcellml/analyserequationast.h#L43-L136).

RLee64 replied:
I think (almost) all types should be added at some point, it's just that I've been gradually going through them to make sure I'm correctly converting to and from these expressions.

The only one I'm not so sure about are piecewise functions, I highly doubt rearranging for those is supported by symengine (considering how symengine is unable to inverse transcedental functions). It doesn't feel like there's much point for us to spend time trying to rearrange something if it's just going to fail.

I also want to do a bit of testing on differential functions to make sure things function as expected.

agarny replied:
That's kind of fine but then you should make it clear in the PR. In fact, a PR should be linked to a GitHub issue that describes what we want/need to do and where we can discuss things prior implementation. From where I am standing, your PR has an unclear objective.

RLee64 replied:
Tried to make an update to the PR descrption, hopefully it's sufficient now?

agarny replied:
Ok, but remember that, in the future, a GitHub issue will first need to be created and then a PR.

agarny commented on src/analyser.cpp:
```diff
@@ -257,29 +309,86 @@ AnalyserEquationAstPtr AnalyserInternalEquation::parseSymEngineExpression(const
+    // All children except the last are to be assigned as left children in the AST tree.
+    for (int i = 0; i + 1 < children.size(); ++i) {
+        auto childSeExpression = children[i];
+        AnalyserEquationAstPtr childAst = parseSymEngineExpression(childSeExpression, currentAst, variableMap);
```
Just have `auto childAst = parseSymEngineExpression(children[i], currentAst, variableMap);`

RLee64 replied:
3a553bed57b4b83352c4fb2bb144944e0672100b

agarny replied:
I would also remove `childSeExpression` which is only used once and is short enough to be inlined.

RLee64 replied:
f50049e79eeb8ac3cfd23ab71e9fadae927dceee

agarny commented on src/analyser.cpp:
```diff
@@ -257,29 +309,86 @@ AnalyserEquationAstPtr AnalyserInternalEquation::parseSymEngineExpression(const
+    // Set the final node as the right child.
+    if (children.size() >= 2) {
+        auto childSeExpression = children.back();
+        AnalyserEquationAstPtr childAst = parseSymEngineExpression(childSeExpression, currentAst, variableMap);
```
Just have `auto childAst = parseSymEngineExpression(children.back(), currentAst, variableMap);`

RLee64 replied:
3a553bed57b4b83352c4fb2bb144944e0672100b

agarny replied:
I would also remove `childSeExpression` which is only used once and is short enough to be inlined.

RLee64 replied:
f50049e79eeb8ac3cfd23ab71e9fadae927dceee

agarny commented on tests/resources/analyser/symengine/unarranged_polynomials.cellml:
```diff
@@ -0,0 +1,66 @@
+      </apply>
+
+      <!-- Equation 3: c^3 + 3*c = 30 -->
+      <!-- Unique solution: c = 3 -->
```
This is wrong. If we want `c = 3` then the equation should read `c^3 + 3*c = 36`.

RLee64 replied:
f8a39f52c927c58f18b0da5dc070f701f1d65f4e

agarny commented on tests/resources/analyser/symengine/unarranged_combination.cellml:
How are these test equations relevant? They already solve for `a`, `b`, `c`, and `d`, respectively.

RLee64 replied:
I haven't gotten around to cleaning those test cases yet.

As a matter of fact, I'll probably delete all test cases I haven't reviewed yet will just add them back when I actually need them.

agarny replied:
Yes, a PR should NOT contain unfinished / unneeded code.

RLee64 replied:
9efff06fc79148bdb0cab40ef8532fbcf268bab2

agarny reviewed (changes requested):
- In your tests, in addition to confirming that a model is of algebraic type, it would be good to confirm that its equation looks the way we would expect. For this, you could use `Generator::equationCode()`.
- No need to have our CellML files named `unarranged_xxx.cellml`. They can simply be named `xxx.cellml` since they are located under the `symengine` folder.
- For the different equations you are testing, it would be nice to have a comment on why they are being tested.

RLee64 committed:
Made explicit types auto

RLee64 committed:
Deleted unused test cases

RLee64 committed:
Renamed symengine cellml test files

RLee64 committed:
Fix issue with recreating unary functions for AST

RLee64 committed:
Cleaned up symengine trigonometric test case

RLee64 committed:
Fixed children not setting their parent AST

I can't believe I didn't get an error for this sooner!!!

RLee64 committed:
Cleaned symengine test cases

- Added comments
- Tried to give each test case a more 'unique' purpose

RLee64 committed:
Test expected generator equation code

Also moved initialising constants to the bottom of the mathml block so we guarantee that our tested equation range always starts at 0

RLee64 reviewed (commented):
> In your tests, in addition to confirming that a model is of algebraic type, it would be good to confirm that its equation looks the way we would expect. For this, you could use Generator::equationCode().

ec0ace2a3a629ea543ca00e2c7c56daf6f6b57d4

> No need to have our CellML files named unarranged_xxx.cellml. They can simply be named xxx.cellml since they are located under the symengine folder.

f8a39f52c927c58f18b0da5dc070f701f1d65f4e

> For the different equations you are testing, it would be nice to have a comment on why they are being tested.

fc1cf1c94c935dfda779e6eb16c2cd50b1ed8824

In addition to these changes I've also done the following
- Cleaned up the trigonometric test case so it's only testing for functionality we know will pass in symengine's current state (i.e. no operations needed to inverse functions). d79d8fc36d48476099d099b6fc5b176602af5865
- Fixed an issue where unary functions weren't being properly converted back (this wasn't picked up since we hadn't encountered symengine outputting unary functions until the proper trigonometric test case) aab7d89379906f057c17c6335b71580c71982d93
- Fixed issue where ast children weren't being assigned their parents (this one I'm just embarrased I hadn't noticed sooner) 9c1e0b8917a00a9789ec44f9dfae7fc87fef7e8b

RLee64 committed:
Made childSeExpression retrieval inline

RLee64 committed:
Fixed spelling error in symengine test files

RLee64 committed:
Removed redundant symengine multiplication test case

RLee64 committed:
Manual simplification of symengine results

x + (-y) now becomes x - y
-1 * y now becomes -y

this also means
x + (-1 * y) becomes x - y

agarny reviewed (approved):
Approving, but... at some point, we will need to ensure that ALL tests pass, incl. coverage tests, memory checks, etc.
